### PR TITLE
✨ feat(NeighborhoodAttention2D and NeighborhoodAttention1D):

### PR DESCRIPTION
**Pull request by lartpang**
Support to remove RPB without side-effects by setting rpb to a buffer tensor.

---

In some special cases, we may not need rpb.

Considering that the implementation of NATTEN depends on rpb, I provide some simple modifications to support this feature.

@alihassanijr

### CODE REVIEW

File: natten/natten1d.py [hunk 1]
```diff
@@ -27,10 +27,14 @@ def __init__(self, dim, kernel_size, num_heads,
...
+        self.with_rpb = with_rpb
 
         self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
-        self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1)))
-        trunc_normal_(self.rpb, std=.02, mean=0., a=-2., b=2.)
+        if with_rpb:
+            self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1)))
+            trunc_normal_(self.rpb, std=.02, mean=0., a=-2., b=2.)
+        else:
+            self.register_buffer('rpb', torch.zeros(num_heads, (2 * kernel_size - 1)))
```
Comment (alihassanijr) [inline natten/natten1d.py:37]:
```suggestion
        self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1)), requires_grad=with_rpb)
        if with_rpb:
            trunc_normal_(self.rpb, std=.02, mean=0., a=-2., b=2.)
```
Could you format it like this instead?

File: natten/natten2d.py [hunk 2]
```diff
@@ -34,10 +34,14 @@ def __init__(self, dim, kernel_size, num_heads,
...
+        self.with_rpb = with_rpb
 
         self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
-        self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1), (2 * kernel_size - 1)))
-        trunc_normal_(self.rpb, std=.02, mean=0., a=-2., b=2.)
+        if with_rpb:
+            self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1), (2 * kernel_size - 1)))
+            trunc_normal_(self.rpb, std=.02, mean=0., a=-2., b=2.)
+        else:
+            self.register_buffer('rpb', torch.zeros(num_heads, (2 * kernel_size - 1), (2 * kernel_size - 1)))
```
Comment (alihassanijr) [inline natten/natten2d.py:44]:
```suggestion
        self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1), (2 * kernel_size - 1)), requires_grad=with_rpb)
        if with_rpb:
            trunc_normal_(self.rpb, std=.02, mean=0., a=-2., b=2.)
```
Same comment here.

**Review by alihassanijr — CHANGES_REQUESTED** [review]
Thanks for the PR again.
We actually might just allow for `None` to be passed to the backend in future versions to just skip over the RPB calls, specifically the grad function that is non-deterministic.

But for now, I think we should be good to merge this.
Could you make the changes I requested though?

**Comment by lartpang** [comment]
@alihassanijr 
Thanks for your reply and the valuable advice.
About the above changes, I'm actually a little confused here.
Besides the more concise form of your proposal, are there some other differences between the two versions?

i.e.:

`self.register_buffer(tensor)` vs `nn.Parameter(tensor, requires_grad=False)`, are they really equivalent??

**Comment by alihassanijr** [comment]
It is just the preferred way of doing it.
Buffers are exactly that, buffers.
Bias terms are never buffers, they are parameters.
Now the "correct" way of doing it would be to set rpb to `None`, similar to standard torch practice; however that would not work until we modify the backend to support NoneTypes for rpb.

**Comment by lartpang** [comment]
> Buffers never get stored into state dict, parameters do. 

The `persistent=True` of `register_buffer` will make the buffer be part of the module’s state_dict.

https://pytorch.org/docs/1.8.0/generated/torch.nn.Module.html?highlight=buffer#torch.nn.Module.register_buffer

File: natten/natten1d.py [hunk 3]
```diff
@@ -27,10 +27,12 @@ def __init__(self, dim, kernel_size, num_heads,
         self.kernel_size = kernel_size
         self.dilation = dilation or 1
         self.window_size = self.kernel_size * self.dilation
+        self.with_rpb = with_rpb
 
         self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
-        self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1)))
-        trunc_normal_(self.rpb, std=.02, mean=0., a=-2., b=2.)
+        self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * kernel_size - 1)), requires_grad=with_rpb)
+        if with_rpb:
```
Comment (alihassanijr) [inline natten/natten1d.py:34]:
```suggestion
        if self.bias:
```

File: natten/natten1d.py [hunk 4]
```diff
@@ -27,10 +27,12 @@ def __init__(self, dim, kernel_size, num_heads,
         self.kernel_size = kernel_size
         self.dilation = dilation or 1
         self.window_size = self.kernel_size * self.dilation
+        self.with_rpb = with_rpb
```
Comment (alihassanijr) [inline natten/natten1d.py:30]:
```suggestion
        self.bias = bias
```

File: natten/natten1d.py [hunk 5]
```diff
@@ -17,7 +17,7 @@ class NeighborhoodAttention1D(nn.Module):
     """
     def __init__(self, dim, kernel_size, num_heads,
                  qkv_bias=True, qk_scale=None, attn_drop=0., proj_drop=0.,
-                 dilation=None):
+                 dilation=None, with_rpb=True):
```
Comment (alihassanijr) [inline natten/natten1d.py:20]:
```suggestion
                 dilation=None, bias=True):
```

File: natten/natten1d.py [hunk 6]
```diff
@@ -57,5 +59,4 @@ def forward(self, x):
         return self.proj_drop(self.proj(x))
 
     def extra_repr(self) -> str:
-        return f'kernel_size={self.kernel_size}, dilation={self.dilation}, head_dim={self.head_dim}, num_heads={self.num_heads}'
-
+        return f'kernel_size={self.kernel_size}, dilation={self.dilation}, head_dim={self.head_dim}, num_heads={self.num_heads}, with_rpb={self.with_rpb}'
```
Comment (alihassanijr) [inline natten/natten1d.py:62]:
```suggestion
        return f'kernel_size={self.kernel_size}, dilation={self.dilation}, head_dim={self.head_dim}, num_heads={self.num_heads}, bias={self.bias}'
```

File: natten/natten2d.py [hunk 7]
```diff
@@ -18,7 +18,7 @@ class NeighborhoodAttention2D(nn.Module):
     """
     def __init__(self, dim, kernel_size, num_heads,
                  qkv_bias=True, qk_scale=None, attn_drop=0., proj_drop=0.,
-                 dilation=None):
+                 dilation=None, with_rpb=True):
```
Comment (alihassanijr) [inline natten/natten2d.py:21]:
```suggestion
                 dilation=None, bias=True):
```

File: natten/natten2d.py [hunk 8]
```diff
@@ -34,10 +34,12 @@ def __init__(self, dim, kernel_size, num_heads,
                 f"Dilation must be greater than or equal to 1, got {dilation}."
             self.dilation = dilation or 1
             self.window_size = self.kernel_size * self.dilation
+        self.with_rpb = with_rpb
```
Comment (alihassanijr) [inline natten/natten2d.py:37]:
```suggestion
        self.bias = bias
```

File: natten/natten2d.py [hunk 9]
```diff
@@ -71,7 +73,7 @@ def forward(self, x):
         return self.proj_drop(self.proj(x))
 
     def extra_repr(self) -> str:
-        return f'kernel_size={self.kernel_size}, dilation={self.dilation}, head_dim={self.head_dim}, num_heads={self.num_heads}'
+        return f'kernel_size={self.kernel_size}, dilation={self.dilation}, head_dim={self.head_dim}, num_heads={self.num_heads}, with_rpb={self.with_rpb}'
```
Comment (alihassanijr) [inline natten/natten2d.py:76]:
```suggestion
        return f'kernel_size={self.kernel_size}, dilation={self.dilation}, head_dim={self.head_dim}, num_heads={self.num_heads}, bias={self.rpb.requires_grad}'
```

**Review by alihassanijr — COMMENTED** [review]
Actually could we refactor `with_rpb` -> `bias` as suggested above?
That would make things easier for future versions.

**Comment by lartpang** [comment]
@alihassanijr 

Following your suggestion, I committed the latest changes. You can check it out.

**Comment by alihassanijr** [comment]
Great. Thank you!

**Pull request merged**